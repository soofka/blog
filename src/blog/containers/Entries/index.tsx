import { reaction } from 'mobx';
import { inject } from 'mobx-react';
import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';

import { LanguageStoreInterface } from 'blog/store/language';

import { EntriesList } from 'blog/components/EntriesList';
import { ErrorBox } from 'blog/components/ErrorBox';
import { Label } from 'blog/components/Label';
import { LoadingCover } from 'blog/components/LoadingCover';
import { EntriesMeta } from 'blog/components/Meta/components/EntriesMeta';

import {
  getEntries,
  RequestSuccessResponseInterface,
  validateRequestSuccessResponse,
} from 'blog/common/helpers';

import { EntryInterface } from 'common/types';

export interface EntriesParamsInterface {
  tag?: string;
  date?: string;
  niceUrl?: string;
}

export interface EntriesPropsInterface extends RouteComponentProps<EntriesParamsInterface> {
  languageStore?: LanguageStoreInterface;
}

interface EntriesStateInterface {
  entries: EntryInterface[];
  loading: boolean;
}

export class EntriesContainer extends React.Component<EntriesPropsInterface, EntriesStateInterface> {
  constructor(props: EntriesPropsInterface) {
    super(props);
    this.state = { entries: undefined, loading: false };

    reaction(
      () => this.props.languageStore.language,
      () => {
        this.getEntries();
      },
    );
  }

  componentDidMount() {
    if (!this.state.entries) {
      this.getEntries();
    }
  }

  getEntries() {
    this.setState({ loading: true });
    getEntries(this.props.languageStore.getLanguage())
      .then((response: RequestSuccessResponseInterface) => {
        const entries = validateRequestSuccessResponse(response, true)
          ? response.data
          : undefined;
        this.setState({ entries, loading: false });
      });
  }

  render() {
    const { match: { params } } = this.props;
    const { entries, loading } = this.state;

    if (loading) {
      return (
        <LoadingCover />
      );
    }

    if (!entries) {
      return (
        <ErrorBox
          message={<Label name="errors__no_entries_in_db" />}
        />
      );
    }

    const entriesFiltered = this.filterEntries(
      prepareEntries(entries),
      params,
    );

    return (
      <div>
        {this.getMeta()}
        <EntriesList
          entries={entriesFiltered}
          fullEntry={this.isEntryFull()}
        />
      </div>
    );
  }

  getMeta(): JSX.Element {
    return <EntriesMeta/>;
  }

  filterEntries(entries: EntryInterface[], params: EntriesParamsInterface): EntryInterface[] {
    return entries;
  }

  isEntryFull(): boolean {
    return false;
  }
}

const prepareEntries = (entries: EntryInterface[]): EntryInterface[] => filterPrivateEntries(entries);

const filterPrivateEntries = (entries: EntryInterface[]): EntryInterface[] =>
  entries.filter((entry: EntryInterface) => entry.public);

export const Entries = inject('languageStore')(EntriesContainer);
