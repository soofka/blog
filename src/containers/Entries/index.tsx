import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';

import LoadingCover from 'components/LoadingCover';
import ErrorBox from 'components/ErrorBox';
import EntriesList from 'components/EntriesList';

import { EntryInterface } from 'containers/Entry';

import AssetsProvider from 'common/AssetsProvider';
import LabelsProvider from 'common/LabelsProvider';
import RequestHandler, { RequestHandlerResponseInterface } from 'common/RequestHandler';

export interface EntriesParamsInterface {
  tag?: string;
  date?: string;
  niceUrl?: string;
}

export interface EntriesPropsInterface extends RouteComponentProps<EntriesParamsInterface> {
  language: string;
}

interface EntriesStateInterface {
  entries: EntryInterface[];
  loading: boolean;
}

export class Entries extends React.Component<EntriesPropsInterface, EntriesStateInterface> {
  constructor(props: EntriesPropsInterface) {
    super(props);
    this.state = { entries: undefined, loading: false };
  }

  componentDidMount() {
    if (!this.state.entries) {
      this.getEntries();
    }
  }

  componentDidUpdate(prevProps: EntriesPropsInterface) {
    if (this.props.language !== prevProps.language) {
      this.getEntries();
    }
  }

  getEntries() {
    this.setState({ loading: true });
    AssetsProvider
      .getEntries(this.props.language)
      .then((response: RequestHandlerResponseInterface) => {
        const entries = RequestHandler.validateResponse(response) ? response.data : undefined;
        this.setState({ entries, loading: false });
      });
  }

  render() {
    const { language, match: { params } } = this.props;
    const { entries, loading } = this.state;

    if (loading) {
      return (
        <LoadingCover />
      );
    }

    if (!entries) {
      return (
        <ErrorBox
          language={language}
          message={LabelsProvider.getLabel('errors__no_entries_in_db', language)}
        />
      );
    }

    const entriesFiltered = this.filterEntries(
      prepareEntries(entries),
      params,
    );

    return (
      <EntriesList
        language={language}
        entries={entriesFiltered}
        fullEntry={this.isEntryFull()}
      />
    );
  }

  filterEntries(entries: EntryInterface[], params: EntriesParamsInterface): EntryInterface[] {
    return entries;
  }

  isEntryFull(): boolean {
    return false;
  }
}

function prepareEntries(entries: EntryInterface[]): EntryInterface[] {
  return filterPrivateEntries(entries);
}

function filterPrivateEntries(entries: EntryInterface[]): EntryInterface[] {
  return entries.filter((entry: EntryInterface) => {
    return entry.isPublic === true;
  });
}

export default Entries;
