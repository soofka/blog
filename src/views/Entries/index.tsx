import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';

import { AssetsProvider, AssetsProviderResponse } from 'common/AssetsProvider';

import LoadingCover from 'components/LoadingCover';
import ErrorBox from 'components/ErrorBox';
import EntriesList from 'components/EntriesList';

import { EntryInterface } from 'containers/Entry';

export interface EntriesParamsInterface {
  tag?: string;
  date?: string;
  niceUrl?: string;
}

export interface EntriesPropsInterface extends RouteComponentProps<EntriesParamsInterface > {}

interface EntriesStateInterface {
  entriesResponse: AssetsProviderResponse;
  isEntryFull: boolean;
}

export class Entries extends React.Component<EntriesPropsInterface, EntriesStateInterface> {
  constructor(props: EntriesPropsInterface) {
    super(props);
    this.state = {
      entriesResponse: null,
      isEntryFull: false,
    };
  }

  componentDidMount() {
    AssetsProvider.getEntries().then((response: AssetsProviderResponse) => {
      this.setState({
        entriesResponse: response,
      });
    });
  }

  render() {
    const { entriesResponse, isEntryFull } = this.state;

    const isValidResponse = entriesResponse !== null;
    const isSuccessResponse = isValidResponse && entriesResponse.status === 200;

    if (!isValidResponse) {
      return (
        <LoadingCover />
      );
    }

    if (!isSuccessResponse) {
      return (
        <ErrorBox
          message="Oops. Entries database crashed. Please try again."
        />
      );
    }

    const entries = this.filterEntries(
      prepareEntries(entriesResponse.data),
      this.props.match.params,
    );

    return (
      <EntriesList
        entries={entries || []}
        isEntryFull={isEntryFull || false}
      />
    );
  }

  filterEntries(entries: EntryInterface[], params: EntriesParamsInterface): EntryInterface[] {
    return entries;
  }
}

function prepareEntries(entries: EntryInterface[]): EntryInterface[] {
  return filterPrivateEntries(entries);
}

function filterPrivateEntries(entries: EntryInterface[]): EntryInterface[] {
  return entries.filter((entry: EntryInterface) => {
    return entry.isPublic === true;
  });
}

export default Entries;
