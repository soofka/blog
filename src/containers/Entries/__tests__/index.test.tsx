import * as React from 'react';
import { shallow } from 'enzyme';

import Entries from 'containers/Entries';

describe('<Entries/>', () => {

  it('renders correctly', () => {
    expect(shallow(<Entries/>)).toMatchSnapshot();
  });

});
