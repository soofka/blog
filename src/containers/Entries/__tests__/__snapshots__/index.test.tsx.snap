exports[`<Entries/> renders correctly 1`] = `<Component />`;
